### PR TITLE
Book finder

### PR DESCRIPTION
**Pull request by riascho**
Merging this exam project exercise from Codecademy. Feel free to have a look. I tried to make the documentation better this time. I only had to write code for the 4 mentioned functions in `script.js`, so everything else was done by codecademy. 
However, I would be curious to know how you would have done things differently. 

For example, I thought that referencing the array of book objects in multiple places seemed a bit much (`books` gets called in both the `filterBooks` and `searchBtnClickHandler` function). I feel like, at least for the last one, it is not necessary. 
However, I can see how this can be used to differentiate the search when maybe another library (e.g. music) is introduced. 

I also tried to avoid shadowing as much as possible this time. But it's hard because you are referencing the same things several times. So I wonder, if you had any advice for that? 

Thank You Much 😇

### CODE REVIEW

File: bookFinder/bookList.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        title: 'Crying in H Mart',
+        author: 'Michelle Zauner',
+        tags: ['memoir', 'nonfiction', 'autobiography']
+    },
+    {
+        title: 'Wild: From Lost to Found on the Pacific Crest Trail',
+        author: 'Cheryl Strayed',
+        tags: ['nonfiction', 'memoir', 'adventure', 'travel']
+    }
+];
```
Comment (MattSchiller) [inline bookFinder/bookList.js:52]: These are some dope books!

File: bookFinder/script.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    const listOfBooks = flattenObjectValuesIntoArray(library);
+    const matchingBooksIndices = listOfBooks
+        .map((book, index) => {
+            if (book.includes(searchString)) {
+                return index;
+            }
+        })
+        .filter((element) => element !== undefined);
+    return matchingBooksIndices.map((item) => library[item]);
+};
```
Comment (MattSchiller) [inline bookFinder/script.js]: So this code is:
- mapping over the array to convert to index if it's "in"
- then filtering based on that determination

Why not just do it together?

If you're finding that you want to do a map & filter together, you're probably gonna want a `reduce`

```suggestion
const filterBooks = (searchString, library) => {
    const listOfBooks = flattenObjectValuesIntoArray(library);
    const matchingBooks = listOfBooks
        .reduce((matchedBooks, book, index) => {
            if (book.includes(searchString)) {
                matchedBooks.push(library[index]);
            }
            
            return matchedBooks;
        }, [])
   return matchingBooks;
};
```

**Review by MattSchiller — APPROVED** [review]

File: bookFinder/script.js (hunk 2 above)
Comment (riascho) [reply bookFinder/script.js]: Oh my god that's genius! I have only used `reduce` with numbers, but I see what you did there! You used the accumulation parameter as an array and initialized it as such in the `reduce` method's 2nd parameter. 🤯 #mindBlown

Comment (ZackMFleischman) [reply bookFinder/script.js]: Building on this, the way to think about reduce is "I want to transform an array into a new object that requires iterating over the array"

I think Matt's code is super clean here, and is a great learning moment around reduce, but I also think your original code is still pretty good. 

In C# there's an array method called "compact" that basically does what your filter was doing, and mapping things to new values or undefined/null and then filtering out the bad values (ie compacting the array) is a perfectly viable strategy and pattern and you'll likely see it happen a lot. 

Most newer developers aren't comfortable with reduce and find it hard to read so they'll lean heavier on the other methods. It's my hope that you'll skip that part of development quickly and get comfy with reduce early. 

Your quality of code is improving very rapidly by the way. Nice work

**Pull request merged**